tests/src/FunctionalJavascript/CorporateContactFormTest.php

```diff
@@ -0,0 +1,320 @@
+    $this->assertFieldsAreMissing($page);
+
+    // Add contact required values.
+    $this->fillCoreContactFields($page);
+
+    $page->pressButton('Save');
+    $assert->pageTextContains('Contact form Test form has been added.');
+
+    // Assert the values are saved.
+    $this->drupalGet('admin/structure/contact/manage/oe_corporate_form');
+    $this->assertFieldsAreMissing($page);
+  }
+
+  /**
+   * Helper to assert field presence.
+   */
+  protected function assertfieldsAreVisible(): void {
+    /** @var \Behat\Mink\Element\DocumentElement $page */
+    $page = $this->getSession()->getPage();
+
+    foreach ($this->fields as $field_name => $value) {
+      $element = $page->findField($field_name);
+      $this->assertNotEmpty($element);
+      $this->assertTrue($element->isVisible());
+    }
+  }
+
+  /**
+   * Helper to assert fields are missing.
+   */
+  protected function assertFieldsAreMissing(): void {
+    /** @var \Behat\Mink\Element\DocumentElement $page */
+    $page = $this->getSession()->getPage();
+
+    foreach ($this->fields as $field_name => $value) {
+      $element = $page->findField($field_name);
+      $this->assertEmpty($element);
+    }
+  }
+
+  /**
+   * Loop through corporate fields and set test values.
+   */
+  protected function fillCoreContactFields(): void {
+    /** @var \Behat\Mink\Element\DocumentElement $page */
+    $page = $this->getSession()->getPage();
+
+    $page->fillField('label', 'Test form');
+    $page->fillField('recipients', 'test@example.com');
+    // Overcome machine name not accessible.
+    $this->getSession()->executeScript('jQuery("#edit-id").val("oe_corporate_form");');
+  }
+
+  /**
+   * Trigger topic add ajax and assert new fields.
+   */
+  protected function assertTopicAddAjax(): void {
+    /** @var \Behat\Mink\WebAssert $assert */
+    $assert = $this->assertSession();
+    /** @var \Behat\Mink\Element\DocumentElement $page */
+    $page = $this->getSession()->getPage();
+
+    $topic_name_key = "corporate_fields[topics_fieldset][group][0][topic_name]";
+    $topic_email_key = "corporate_fields[topics_fieldset][group][0][topic_email_address]";
+
+    $page->fillField($topic_name_key, $this->fields[$topic_name_key]);
+    $page->fillField($topic_email_key, $this->fields[$topic_email_key]);
+
+    // Test add topic group.
+    $add_topic = $page->find('css', 'input[value="Add topic"]');
+    $this->assertNotEmpty($add_topic);
+    $add_topic->click();
+    $assert->assertWaitOnAjaxRequest();
+
+    $topic_name_key = "corporate_fields[topics_fieldset][group][1][topic_name]";
+    $topic_email_key = "corporate_fields[topics_fieldset][group][1][topic_email_address]";
+
+    $element = $page->findField($topic_name_key);
+    $this->assertNotEmpty($element);
+    $element = $page->findField($topic_email_key);
+    $this->assertNotEmpty($element);
+  }
+
+  /**
+   * Trigger topic remove ajax and assert no new fields.
+   */
+  protected function assertTopicRemoveAjax(): void {
```

Comment:
This method is problematic. It can only be run after `assertTopicAddAjax()` which makes it tied to add. I suggest you do one single method that tests the topic ajax. And does it more thoroughly. Adding values, removing values. And also more than one set of values. And validation of the values (to not be able to add more if incorrect or empty)

This also needs to be tested after saving and coming back for edit.

Reply:
done